### PR TITLE
fix: Canvas card の identity を「ワイヤの綴り」から「ファイル」にする (#1976 PR 1/2)

### PR DESCRIPTION
**Pull request by isamu**
## Summary

#1976 の**1 本目**。issue のコメントで決めた「軽い版」の PR 1 — **card identity だけ**を直す。
`storyWirePath` のゲート（＝ボタンが出る条件）は触らないので、**この PR では issue を閉じない**。

card identity は `filePathIdentity` が決めていて、**ワイヤの綴りそのもの**（`root\0filePath`）だった。
ワイヤの綴りはファイルの性質ではなく「**このサーバが何を登録しているか**」の関数なので、
#1976 とは独立に今日踏める欠陥が 2 つある（どちらも純関数で再現を取ってから直した）:

- **(B) preset を 1 つ足して再起動すると、同じファイルの identity が変わる**
  `file = /Users/me/w/proj/decks/x.json` は、workspace だけ登録なら `W\0stories/proj/decks/x.json`、
  `…/proj` も登録すると `P\0stories/decks/x.json`。前後に作られたカードは 1 つのデッキで 2 枚になり、
  新しい方が古い方を supersede しない。
- **(C) default root のカードにはパス成分が無い**
  `/w1/artifacts/stories/x.json` も `/w2/artifacts/stories/x.json` も identity は `stories/x.json`。
  #1933 が named root について入れた守りが、default root には効いていなかった。

**直し方**: カードが持つ情報から canonical 絶対パスを組み立て、それを identity にする。
root id → canonical ディレクトリ、tail はどちらの綴りでも同じなので、上の 2 つの綴りは
どちらも `/Users/me/w/proj/decks/x.json` に解決する。絶対パスの `filePath` も同じ規則で読むので、
**「root 経由で開いたカード」と「絶対パスで開いたカード」が 1 枚に畳まれる** — これが PR 2
（ゲートを緩めてボタンを出す）の前提。解決できないカードは**旧来の文字列にフォールバック**するので、
マイグレーションも二重キーも要らない。

順番を逆にできない理由: ボタンを先に出すと割れたカードが生まれ、それが永続 store
（`~/.mulmoterminal/toolresults`）に残る。あとから identity を変えるとき、移行対象が
「設計上の互換性」ではなく「実際に割れて保存された履歴」になる。

設計メモ: [`plans/fix-1976-canonical-card-identity.md`](plans/fix-1976-canonical-card-identity.md)

## Items to Confirm / Review

1. **`/api/config` に `canonical` を足した**（`storiesRoots[].canonical`）。issue のコメントは
   「`paths` に canonical も入っているから使える」としていたが、**どれが canonical かの印が無く、
   位置も一定しない**（symlink preset がマージされる経路で順序が変わる）。実サーバで確認した実測:
   symlink 経由で起動すると `paths[0]` は **launcher で打った symlink 綴り**で、canonical は 2 番目。
   ブラウザは realpath できないので、どの綴りが解決済みかはサーバしか言えない。
   → **ラベルにする判断でよいか**（位置の規約にする案もあった）。
2. **`identityOf` の signature を `(result, storyRoots)` に変えた**。root 表は mulmoScript 固有だが
   全 accessor に渡している（引数 1 つだけ宣言している accessor はそのままで assignable）。
   `IdentityContext` のような箱を噛ませるほどではない、という判断。
3. **限界（今回入れていないこと）**: プラグインは**絶対パスの `filePath` を呼び出し側が綴ったまま返す**
   （core の `locate` が `byPath` にその文字列を渡す）。なので **symlink 経由の綴りで開いた絶対パスは
   畳めない** — 実サーバで再現も確認済み。塞ぐには `resolveStory` の realpath 済み `absolutePath` を
   card payload に載せる「正確な版」が要る。両側とも値は手元にあるが、この PR には入れていない。
4. **html の絶対パスも `dirPathKey` で正規化されるようになった**（`/a/./b.html` と `/a/b.html` が 1 枚）。
   意図した副作用。別ファイルが混ざることはないが、想定どおりか。

5. **レビューで足した規則（Windows）**: card path が「サーバの *現在のドライブ* に依存する綴り」の
   ときは identity を作らず、旧来の文字列に落ちる。`\deck.json` はどの host でもそう扱い、
   `/deck.json` は **サーバが自分の root をドライブ付き / UNC で綴っている場合だけ**そう扱う
   （POSIX ではそれが唯一の絶対形なので、一律に拒めば機能が死ぬ）。host の判定は `navigator` では
   なく **サーバが登録した root の綴り**から取る — この画面はスマホで開かれていることがある。

## 検証

- **(B)(C) を現行コードに対して再現させてから**直した。回帰テストは Files ペインの実チェーン
  （`storyWirePath` → card → `filePathIdentity`）を通す（片側だけでは欠陥が見えないため）。
- 解決を無効化して**全部赤になること**を確認（identity 10 本 + パネル 2 本）。
- `yarn format` / `lint` / `typecheck` / `build` / `test`（12099 passed）ローカル green。
- **実サーバで確認**（workspace を symlink 経由で起動、preset に `ws/proj` を登録）:
  - `/api/config` の `canonical` が resolved 綴りで、`paths[0]` ではない。
  - プラグインの実レスポンス 4 通り（default root / root+tail / 絶対パス canonical / 絶対パス symlink）を
    identity にかけると、`root+tail` と `絶対パス canonical` が 1 つに畳まれる。symlink 綴りだけ別（上記 3）。
  - preset 追加前後の 2 つのワイヤ綴りが同じ identity に解決される（(B) の実機再現）。
  - 実際に保存された 2 枚の toolResult を実 config で解決 → 1 identity。
  - ビルド済みアプリをブラウザで読み込み: 描画される / uncaught console error なし。
  - **やっていないこと**: Canvas ペインを実ブラウザで開いてカードを目視するところまでは自動化できなかった
    （シェルセルの UI 操作が不安定）。表示の畳み込み自体は **GuiPanel を実コンポーネントとしてマウントする**
    spec で固定してある（config 到着で再 collapse されることも含む）。

## User Prompt

- https://github.com/receptron/mulmoterminal/issues/1976 のコメントを確認して、その方法で進める

Refs #1976

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01XD523v7QBEqiYymHgLYbVT

work in mulmoterminal4

## Summary by CodeRabbit

- **Bug Fixes**
  - Card identity now consistently uses canonical absolute paths, preventing duplicates from different path spellings.
  - Decks with identical names in different story roots remain distinct.
  - POSIX, Windows, and UNC rooted paths are recognized consistently; single-leading-backslash paths are not treated as rooted.
  - Compatibility is preserved when story-root configuration is unavailable or incomplete.
  - Story-root configuration now retains resolved canonical paths.

- **Tests**
  - Added coverage for canonical paths, multiple roots, symlinks, cross-platform paths, and delayed configuration loading.

### CODE REVIEW

File: src/utils/canvasCardPath.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+export function canonicalCardPath(filePath: string, root: string | null, dirs: StoryRootDirs): string | null {
+  // Already the answer, whoever minted it. Keyed all the same, so `/w/./x.json` and `/w/x.json` are
+  // one card.
+  if (isRootedPath(filePath)) return dirPathKey(filePath);
+  if (!filePath.startsWith(STORY_WIRE_PREFIX)) return null;
+  const tail = filePath.slice(STORY_WIRE_PREFIX.length);
+  const base = root === null ? defaultStoriesDir(dirs.workspace) : (dirs.byId[root] ?? null);
+  // A bare `stories/` names the directory, not a card in it.
+  if (base === null || tail === "") return null;
+  return dirPathKey(`${base}/${tail}`);
```
[the remark this answers is not in the document]
Comment (isamu) [reply src/utils/canvasCardPath.ts]: Disagreeing — the prefix **is** stripped, on the line above the one this points at.

```ts
// src/utils/canvasCardPath.ts:74-79
if (!filePath.startsWith(STORY_WIRE_PREFIX)) return null;   // STORY_WIRE_PREFIX = "stories/"
const tail = filePath.slice(STORY_WIRE_PREFIX.length);      // "stories/decks/x.json" -> "decks/x.json"
const base = root === null ? defaultStoriesDir(dirs.workspace) : (dirs.byId[root] ?? null);
if (base === null || tail === "") return null;
return dirPathKey(`${base}/${tail}`);                        // "/w/proj" + "/decks/x.json"
```

`tail` is the path with `stories/` already removed, so a named root joins `<root canonical>/<tail>` and the default root joins `<workspace>/artifacts/stories/<tail>`. That is exactly the split the finding asks for.

Evidence that it is not `/w/proj/stories/decks/x.json`:

1. `test/src/utils/canvasCardPath.spec.ts:43` asserts `canonicalCardPath("stories/decks/x.json", "P", dirs) === "/Users/me/w/proj/decks/x.json"` and passes — it would fail under the behaviour described.
2. Measured against a running server (workspace launched through a symlink, `ws/proj` registered as a preset): the plugin answers `{filePath: "stories/decks/x.json", root: "dd07cc51e6d7d3e2"}` for the deck at `<ws>/proj/decks/x.json`, and this resolves it to `<ws>/proj/decks/x.json` — the same string the absolute-path card carries, i.e. the two collapse.
3. Same run, the (B) case end to end: `{W, "stories/proj/decks/x.json"}` (before the preset) and `{P, "stories/decks/x.json"}` (after it) both resolve to `<ws>/proj/decks/x.json`.

If the finding was aimed at something else — say the default-root branch keeping `artifacts/stories` — say which input, and I will re-check.

**Comment by isamu** [comment]
## Review ledger — after iteration 1

Findings raised so far and what happened to each. Re-flagging a `FIXED` or
`REJECTED` row is welcome **if the resolution is wrong** — say which row and what
is wrong with it.

| # | iter | bot | finding (one line) | disposition | why / where |
|---|------|-----|--------------------|-------------|-------------|
| 1 | 1 | codex | `canonicalCardPath` joins the wire path without stripping `stories/` for named roots | REJECTED | it is stripped at `src/utils/canvasCardPath.ts:75`, one line above the anchor the finding points at; `test/src/utils/canvasCardPath.spec.ts:43` asserts `/Users/me/w/proj/decks/x.json` and passes; the same case was measured against a running server (real `/api/config` + real plugin payloads) — [full reply](https://github.com/receptron/mulmoterminal/pull/1990#discussion_r3949583461) |
| 2 | 1 | CodeRabbit | MD040: the plan's two example fences carry no language tag | FIXED | commit ceeaa20 |

CI on `ab0b2ee`: all checks green (lint-and-build macOS + Ubuntu, test_windows,
package-smoke, knip, jscpd, Socket).

File: common/dirPathKey.ts [hunk 2]
```diff
@@ -43,3 +43,8 @@ export function dirPathKey(path: string): string {
 
 /** Whether two paths name the same directory, as far as spelling can tell. */
 export const isSameDirPath = (a: string | null | undefined, b: string | null | undefined): boolean => !!a && !!b && dirPathKey(a) === dirPathKey(b);
+
+/** Whether `path` names a place from a root — POSIX `/…`, a Windows drive root, a UNC share —
+ *  rather than one relative to a directory the spelling does not name. The same three spellings
+ *  `dirPathKey` keeps, so a caller cannot disagree with it about what "absolute" means. */
+export const isRootedPath = (path: string): boolean => rootOf(path.trim()) !== "";
```
[the remark this answers is not in the document]
Comment (isamu) [reply common/dirPathKey.ts]: Fixed in `6377b34f` — confirmed first, then narrowed.

Measured before the change:

```
"\deck.json"            rooted: true   key: "/deck.json"
"C:\deck.json"          rooted: true   key: "C:/deck.json"
"/deck.json"            rooted: true   key: "/deck.json"
```

so the split and the conflation are both real, exactly as described.

`isRootedPath` now accepts only the three spellings that name a file on their own — `/…`, `X:/` or `X:\`, and a UNC share — and refuses a single leading backslash, so such a card keeps the identity it had before this PR rather than being keyed into something a drive-qualified card cannot match.

`dirPathKey` itself is unchanged: folding `\deck.json` to `/deck.json` is right for its job (comparing two spellings of a directory) and the two now differ only here, which the doc comment says. That is the same distinction `rootOf` already draws for `C:foo` and `\server\share`.

Windows boundary tests added where you asked, plus one at each caller:

- `test/common/dirPathKey.spec.ts` — the three rooted spellings, the drive-relative `C:foo`, and `\deck.json` / `\work\deck.json` refused while `dirPathKey("\deck.json")` still keys to `/deck.json`.
- `test/src/utils/canvasCardPath.spec.ts` — `\decks\x.json` resolves to nothing; `C:\decks\x.json` resolves to `C:/decks/x.json`.
- `test/src/canvasIdentity.spec.ts` — a card carrying `\decks\x.json` keeps its legacy identity.

**Comment by isamu** [comment]
## Review ledger — after iteration 2

Findings raised so far and what happened to each. Re-flagging a `FIXED` or
`REJECTED` row is welcome **if the resolution is wrong** — say which row and what
is wrong with it.

| # | iter | bot | finding (one line) | disposition | why / where |
|---|------|-----|--------------------|-------------|-------------|
| 1 | 1 | codex | `canonicalCardPath` joins the wire path without stripping `stories/` for named roots | REJECTED | it is stripped at `src/utils/canvasCardPath.ts:75`, one line above the anchor; `test/src/utils/canvasCardPath.spec.ts:43` asserts `/Users/me/w/proj/decks/x.json` and passes; the same case measured against a running server — [reply](https://github.com/receptron/mulmoterminal/pull/1990#discussion_r3949583461) |
| 2 | 1 | CodeRabbit | MD040: the plan's two example fences carry no language tag | FIXED | commit ceeaa20 |
| 3 | 2 | codex | `isRootedPath` accepts a single leading backslash, whose drive is unknown on Windows — identity splits from `C:\…` and collides with POSIX `/…` | FIXED | measured (`rooted: true`, key `/deck.json`), narrowed to the three self-naming spellings, such a card keeps its legacy identity; commit 6377b34 with Windows boundary tests in `dirPathKey.spec.ts` + both callers — [reply](https://github.com/receptron/mulmoterminal/pull/1990#discussion_r3949651342) |

`yarn format` / `lint` / `typecheck` / `test` (12106 passed) green locally on 6377b34.

File: common/dirPathKey.ts [hunk 3]
```diff
@@ -43,3 +43,22 @@ export function dirPathKey(path: string): string {
...
+ * it would give `\deck.json` and `C:\deck.json` two identities for one file while conflating the
+ * first with a POSIX `/deck.json` (Codex P2 on #1976). A caller that needs a decision here gets
+ * "no", and whatever it falls back to is at worst what it did before.
+ *
+ * The same distinction `rootOf` already draws for `C:foo` and `\server\share`: a spelling that
+ * means something else must not borrow a rooted one's key.
+ */
+export const isRootedPath = (path: string): boolean => {
+  const trimmed = path.trim();
+  return /^[a-zA-Z]:[/\\]/.test(trimmed) || /^[/\\]{2}/.test(trimmed) || trimmed.startsWith("/");
```
[the remark this answers is not in the document]
Comment (isamu) [reply common/dirPathKey.ts:63]: Fixed in `eb72693b`, taking the first of the two options you named.

`isRootedPath` cannot refuse `/deck.json` — on every non-Windows host that is THE absolute spelling, and refusing it would leave the whole feature with no identity to build. So the decision moved to `canonicalCardPath`, which has the one signal a browser actually has: the roots the server registered. If those are drive-qualified, a card path that is neither drive-qualified nor a UNC share resolves to nothing and keeps its legacy identity.

Read off the roots rather than `navigator`, deliberately: this page can be open on a phone while the files are on a Windows machine, so the browser's own platform says nothing about theirs.

Boundary tests added in `test/src/utils/canvasCardPath.spec.ts` ("says nothing for a drive-less path when the server's roots are Windows ones"):

| card path | roots `C:/Users/me/w` | roots `/Users/me/w` |
|---|---|---|
| `/decks/x.json` | null | `/decks/x.json` |
| `\decks\x.json` | null | null (previous commit) |
| `C:\decks\x.json` | `C:/decks/x.json` | — |
| `//server/share/x.json` | `//server/share/x.json` | — |
| `stories/decks/x.json` + root `W` | `C:/Users/me/w/decks/x.json` | unaffected |

Root-relative cards are unaffected either way — their base is the root's own spelling.

The residual is the one the PR body already lists as deferred: the plugin echoes an absolute `filePath` as the caller spelled it, so a symlinked spelling still stands apart. The proper close for both is the server putting its resolved path on the card payload, which both sides already compute.

**Comment by isamu** [comment]
## Review ledger — after iteration 3

| # | iter | bot | finding (one line) | disposition | why / where |
|---|------|-----|--------------------|-------------|-------------|
| 1 | 1 | codex | `canonicalCardPath` joins the wire path without stripping `stories/` for named roots | REJECTED | it is stripped at `src/utils/canvasCardPath.ts` one line above the anchor; the spec asserting the resolved path passes; measured against a running server — [reply](https://github.com/receptron/mulmoterminal/pull/1990#discussion_r3949583461) |
| 2 | 1 | CodeRabbit | MD040: the plan's two example fences carry no language tag | FIXED | commit ceeaa20 |
| 3 | 2 | codex | `isRootedPath` accepts a single leading backslash, whose drive is unknown on Windows | FIXED | measured, then narrowed to the three self-naming spellings; such a card keeps its legacy identity — commit 6377b34, Windows boundary tests in `dirPathKey.spec.ts` + both callers |
| 4 | 3 | codex | `/deck.json` has the same Windows meaning, so the same split/collision remains | FIXED | `isRootedPath` cannot refuse it (it is the absolute form everywhere else), so the decision moved to `canonicalCardPath`, which refuses a drive-less path when the server's REGISTERED roots are drive-qualified — commit eb72693, table of boundary cases in the [reply](https://github.com/receptron/mulmoterminal/pull/1990#discussion_r3949711880) |

`yarn format` / `lint` / `typecheck` / `test` (12107 passed) green locally on eb72693.

Still deliberately out of scope, and stated in the PR body: the plugin echoes an
absolute `filePath` as the caller spelled it, so a symlinked spelling stands
apart. The close for that — and for the drive-less cases above — is the server
putting its own resolved path on the card payload.

File: src/utils/canvasCardPath.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+/**
+ * Whether this SERVER spells its own directories with a drive letter.
+ *
+ * Read off the roots it registered, never off `navigator`: the page can be open on a phone while
+ * the files live on a Windows machine, so the browser's own platform says nothing about theirs.
+ * Unknown (no roots yet) reads as POSIX, which is the spelling every other host uses.
+ */
+const windowsSpelled = (dirs: StoryRootDirs): boolean => {
+  const dir = dirs.workspace ?? Object.values(dirs.byId)[0];
+  return dir !== undefined && /^[a-zA-Z]:[/\\]/.test(dir);
```
[the remark this answers is not in the document]
Comment (isamu) [reply src/utils/canvasCardPath.ts]: Fixed in `64f96c6a` — and since this is the third round on one theme, the rule is now stated once instead of another branch being added.

```ts
const DRIVE_OR_SHARE = /^([a-zA-Z]:[/\\]|[/\\]{2})/;

const namesOneFileOnThisServer = (filePath: string, dirs: StoryRootDirs): boolean => {
  if (!isRootedPath(filePath)) return false;
  const serverDir = dirs.workspace ?? Object.values(dirs.byId)[0];
  const windowsSpelled = serverDir !== undefined && DRIVE_OR_SHARE.test(serverDir);
  return !windowsSpelled || DRIVE_OR_SHARE.test(filePath.trim());
};
```

The predicate that recognises a qualified CARD path is the same one applied to a directory the SERVER resolved, so a UNC-rooted host is a Windows host by construction — there is no second list to keep in step.

Boundary cases added (`test/src/utils/canvasCardPath.spec.ts`), server rooted at `//server/share/project`:

| card path | result |
|---|---|
| `/decks/x.json` | null |
| `\decks\x.json` | null |
| `//server/share/project/decks/x.json` | `//server/share/project/decks/x.json` |
| `C:\decks\x.json` | `C:/decks/x.json` |
| `stories/decks/x.json` + root `W` | `//server/share/project/decks/x.json` |

Plus one for reading the host's spelling off a named root when the workspace entry is not there yet, since that is the other way `StoryRootDirs` can be populated.

The residual is unchanged and stated in the PR body: an absolute `filePath` is echoed by the plugin as the caller spelled it, so a symlinked spelling stands apart. Putting the server's resolved path on the card payload closes that and every drive case above; it is deliberately not in this PR.

**Comment by isamu** [comment]
## Review ledger — after iteration 4

| # | iter | bot | finding (one line) | disposition | why / where |
|---|------|-----|--------------------|-------------|-------------|
| 1 | 1 | codex | `canonicalCardPath` joins the wire path without stripping `stories/` for named roots | REJECTED | it is stripped one line above the anchor; the spec asserting the resolved path passes; measured against a running server — [reply](https://github.com/receptron/mulmoterminal/pull/1990#discussion_r3949583461) |
| 2 | 1 | CodeRabbit | MD040: the plan's two example fences carry no language tag | FIXED | commit ceeaa20 |
| 3 | 2 | codex | `isRootedPath` accepts a single leading backslash, whose drive is unknown on Windows | FIXED | commit 6377b34 — narrowed to the spellings that name a file on their own |
| 4 | 3 | codex | `/deck.json` has the same Windows meaning, so the split/collision remains | FIXED | commit eb72693 — decided from the server's own root spelling, since the form cannot be refused everywhere |
| 5 | 4 | codex | a UNC-rooted Windows server is read as POSIX, so `/deck.json` is keyed there | FIXED | commit 64f96c6 — the rule stated once (`namesOneFileOnThisServer`), so the host test IS the qualified-path test — [reply](https://github.com/receptron/mulmoterminal/pull/1990#discussion_r3949788202) |

3, 4 and 5 are one theme; the last commit replaces the enumeration with the rule,
which is what should stop it recurring.

`format` / `lint` / `typecheck` / `test` green locally on 64f96c6. One unrelated
file (`listCollectionProjects.spec.ts`) failed once in the full run at load
average 57 and passes standalone (14/14) — the load, not the change.

**CodeRabbit has been rate limited since iteration 1**, so it has not reviewed
2, 3 or 4. That is recorded rather than counted as a sign-off.

**Comment by isamu** [comment]
## Checkpoint — round 5

Read whole (`git diff origin/main...HEAD`), not as a delta, before reacting to another round.

**Does the diff still do what the title says?** Yes. One rule changed —
`filePathIdentity` compares the FILE a card is about instead of the wire string —
plus the two things that rule needs to exist: `canonical` on `/api/config`'s
stories roots, and the root table reaching `identityOf` through the panel. No
behaviour outside the Canvas's display collapsing is touched, which is what the
body claims.

**Original work vs review response.** 5 commits: the change (ab0b2ee) and four
responses. In lines, the responses are **145 of 672** — and 94 of those 145 are
tests. The only production code they moved is `isRootedPath` and the four lines
in `canonicalCardPath` that decide whether a path names a file on THIS server.
The centre of gravity has not moved.

**Is anything here only because a bot asked?** `namesOneFileOnThisServer` is,
and it stays: without it the identity claims to be a file and is not on Windows.
Its first two versions were an enumeration (`\x`, then `/x`, then UNC) and that
IS the shape of a finding that keeps coming back — 64f96c6 replaced it with the
rule, which is why it should stop. Nothing else was added on request.

**Would I approve this fresh today?** Yes, with the two limits the body already
states: identity is lexical, so a symlink below a root still splits, and the
plugin echoes an absolute `filePath` as the caller spelled it. Both close the
same way — the server putting its resolved path on the card payload — and both
are named rather than hidden.

**Largest remaining risk.** Not the Windows spellings; it is that this is
display-only collapsing and the toolResults store keeps every card, so a wrong
identity hides a card rather than losing one, and a fallback (unregistered root,
config not yet arrived, drive-less path) restores exactly the previous
behaviour. The verification gap I have not closed is the one the body names:
the Canvas pane opening in a real browser was not automated, so the collapse is
pinned by a spec that mounts the real `GuiPanel`, not by a screenshot.

**Not a scope change**, so the clean-round count is unaffected — there is
nothing here that would invalidate an earlier sign-off.

@coderabbitai full review — you have been rate limited since iteration 1 and
have not seen 6377b34, eb72693 or 64f96c6.

File: src/utils/canvasCardPath.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * Which host that is comes from the way the SERVER spells its own registered roots — the same test,
+ * applied to a directory it resolved — and never from `navigator`: this page can be open on a phone
+ * while the files live on Windows. No roots yet reads as POSIX, the spelling every other host uses.
+ */
+const namesOneFileOnThisServer = (filePath: string, dirs: StoryRootDirs): boolean => {
+  if (!isRootedPath(filePath)) return false;
+  const serverDir = dirs.workspace ?? Object.values(dirs.byId)[0];
+  const windowsSpelled = serverDir !== undefined && DRIVE_OR_SHARE.test(serverDir);
+  return !windowsSpelled || DRIVE_OR_SHARE.test(filePath.trim());
```
[the remark this answers is not in the document]
Comment (isamu) [reply src/utils/canvasCardPath.ts]: Fixed — taking the first option you named (`//` preserved only when the host is the Windows one), and the rule is now written once for both directions instead of a Windows-only guard:

```ts
const absoluteOnThisServer = (filePath: string, dirs: StoryRootDirs): string | null => {
  const spelling = filePath.trim();
  const serverDir = dirs.workspace ?? Object.values(dirs.byId)[0];
  if (serverDir !== undefined && DRIVE_OR_SHARE.test(serverDir)) {
    return DRIVE_OR_SHARE.test(spelling) ? dirPathKey(spelling) : null;   // Windows
  }
  return spelling.startsWith("/") ? dirPathKey(spelling.replace(/^\/+/, "/")) : null;  // everywhere else
};
```

Confirmed on the machine first, rather than from the docs — `fs.realpathSync` on this host:

```
"/private/tmp/.../keynote.json"   -> /private/tmp/.../keynote.json
"//private/tmp/.../keynote.json"  -> /private/tmp/.../keynote.json
```

| card path | POSIX-rooted server | Windows-rooted server |
|---|---|---|
| `/Users/me/decks/x.json` | `/Users/me/decks/x.json` | null |
| `//Users/me/decks/x.json` | `/Users/me/decks/x.json` | (a share elsewhere) |
| `///Users/me/decks/x.json` | `/Users/me/decks/x.json` | — |
| `//server/share/x.json` | folds to `/server/share/x.json` | `//server/share/x.json` |
| `C:\decks\x.json` | null (names nothing there) | `C:/decks/x.json` |
| `\decks\x.json` | null | null |

**And one instance of this that no bot flagged, found while making the change**: the ROOT-RELATIVE branch keyed with `dirPathKey` directly, so a workspace at the filesystem root built `//artifacts/stories/x.json` — a UNC root — and that card would not have collapsed with the absolute card for the same file. Both branches now key through `absoluteOnThisServer`, so they cannot disagree. A workspace at `/` is a case this file already takes seriously (Codex P1 on #1934), and there is a test for it now.

The class that is still deferred, and now named in the header: a symlink nobody resolved, and a path differing only in CASE on a case-insensitive volume — the browser cannot decide either, and folding case where the volume IS case-sensitive would merge two real files, which is the worse error. All of them close the same way: the server putting its own resolved path on the card payload.

**Comment by isamu** [comment]
## Review ledger — after iteration 6

| # | iter | bot | finding (one line) | disposition | why / where |
|---|------|-----|--------------------|-------------|-------------|
| 1 | 1 | codex | `canonicalCardPath` joins the wire path without stripping `stories/` | REJECTED | stripped one line above the anchor; spec + a run against a real server |
| 2 | 1 | CodeRabbit | MD040 on the plan's fences | FIXED | ceeaa20 |
| 3 | 2 | codex | `\deck.json` — drive unknown on Windows | FIXED | 6377b34 |
| 4 | 3 | codex | `/deck.json` — same on a Windows host | FIXED | eb72693 |
| 5 | 4 | codex | UNC-rooted Windows server read as POSIX | FIXED | 64f96c6 — rule stated once |
| 6 | 5 | codex | **LGTM** on `64f96c6a` | — | first clean verdict |
| 7 | 5 (re-run, same head) | codex | POSIX `//x` keeps a UNC identity though it is the same file as `/x` | FIXED | c385fbe — measured with `realpath` first; the rule now written in BOTH directions |
| 8 | 6 | *(mine)* | the root-relative branch keyed with `dirPathKey` directly, so a workspace at `/` built `//artifacts/stories/…` and would not collapse with the absolute card | FIXED | c385fbe — observed reading it back, not flagged by a bot; both branches key through one rule |

Clean-round count: **0**. The LGTM at row 6 was followed by a new finding on the
same head, so it does not start the pair.

`format` / `lint` / `typecheck` / `test` (12113 passed) green locally on c385fbe.

Reviewer coverage, since it belongs in the record: **CodeRabbit has been rate
limited since iteration 1** and has seen only `ab0b2ee`. Codex has seen every head.

**Comment by isamu** [comment]
## Iteration 7 — Codex LGTM on `c385fbe`

First clean verdict since the findings started. Per the loop's own rule that is
**one** clean round, not two: commits were pushed in it, so the pair has to be
completed by a round that reviews a head the first one has already seen.

`a9d6e14` (docs only) records the spelling rule in the design note — the table of
which spellings name a file on a POSIX- vs Windows-spelled server, what `null`
means, and the family left open. After its round I will re-trigger on the
unchanged head rather than push again.

**CodeRabbit still has not reviewed anything past `ab0b2ee`** (rate limited since
iteration 1). Its silence is recorded, not counted.

**Comment by isamu** [comment]
## Converged — two clean rounds on the same head

`a9d6e146` was reviewed twice with nothing pushed in between (12:56 and 13:05),
LGTM both times, and every CI check is green.

### Final ledger

| # | iter | bot | finding | disposition |
|---|------|-----|---------|-------------|
| 1 | 1 | codex | `canonicalCardPath` joins the wire path without stripping `stories/` | **REJECTED** — stripped one line above the anchor; spec + a run against a real server |
| 2 | 1 | CodeRabbit | MD040 on the plan's fences | FIXED (ceeaa20) |
| 3 | 2 | codex | `\deck.json` — drive unknown on Windows | FIXED (6377b34) |
| 4 | 3 | codex | `/deck.json` — same meaning on a Windows host | FIXED (eb72693) |
| 5 | 4 | codex | UNC-rooted Windows server read as POSIX | FIXED (64f96c6) |
| 6 | 5 | codex | LGTM, then a new finding on the same head | — |
| 7 | 5b | codex | POSIX `//x` keeps a UNC identity though it is the same file as `/x` | FIXED (c385fbe) |
| 8 | 6 | *(mine)* | the root-relative branch keyed directly, so a workspace at `/` built `//artifacts/stories/…` and would not collapse with the absolute card | FIXED (c385fbe) — not flagged by any bot |
| 9 | 7, 8, 8b | codex | LGTM ×3 | — |

Rows 3–5 and 7 are one theme. c385fbe replaced the enumeration with the rule
(`absoluteOnThisServer`), written in both directions, and a9d6e14 put the
resulting table in the design note.

### Reviewer coverage — please read this before merging

**CodeRabbit has been rate limited since iteration 1 and has reviewed only
`ab0b2ee`.** It has not seen the four commits that followed, including every
Windows/POSIX spelling change. Codex has seen every head. That is a gap in the
record, not a sign-off.

### Verification

- `format` / `lint` / `typecheck` / `test` (12113) green locally on the final head;
  all 11 CI checks green.
- The rule's own boundary cases are pinned in `canvasCardPath.spec.ts` and
  `dirPathKey.spec.ts`; the panel's use of it in `GuiPanelCollapse.spec.ts`.
- Measured against a running server: `/api/config` carries `canonical`, the two
  wire spellings of one deck resolve to one identity, and the two stored
  toolResults for one file resolve to one identity.
- Not automated: opening the Canvas pane in a real browser for THIS PR. (#1992,
  stacked on this branch, does drive the browser end to end and its card is built
  by this rule.)

Ready for a human decision — I am not merging it.
